File: db4on/test/src/com/db4o/db4ounit/common/acid/CrashSimulatingTestCase.cs
namespace com.db4o.db4ounit.common.acid
{
	public class CrashSimulatingTestCase : Db4oUnit.TestCase, Db4oUnit.Extensions.Fixtures.OptOutCS
	{
		public string _name;

		public com.db4o.db4ounit.common.acid.CrashSimulatingTestCase _next;

		internal const bool LOG = false;

		public CrashSimulatingTestCase()
		{
		}

		public CrashSimulatingTestCase(com.db4o.db4ounit.common.acid.CrashSimulatingTestCase
			 next_, string name)
		{
			_next = next_;
			_name = name;
		}

		private bool HasLockFileThread()
		{
			if (!com.db4o.@internal.Platform4.HasLockFileThread())
			{
				return false;
			}
			return !com.db4o.@internal.Platform4.HasNio();
		}

		public virtual void Test()
		{
			if (HasLockFileThread())
			{
				j4o.lang.JavaSystem.Out.WriteLine("CrashSimulatingTestCase is ignored on platforms with lock file thread."
					);
				return;
			}
			string path = System.IO.Path.Combine(System.IO.Path.GetTempPath(), "crashSimulate"
				);
			string fileName = System.IO.Path.Combine(path, "cs");
			com.db4o.foundation.io.File4.Delete(fileName);
			System.IO.Directory.CreateDirectory(path);
			com.db4o.Db4o.Configure().BTreeNodeSize(4);
			CreateFile(fileName);
			com.db4o.db4ounit.common.acid.CrashSimulatingIoAdapter adapterFactory = new com.db4o.db4ounit.common.acid.CrashSimulatingIoAdapter
				(new com.db4o.io.RandomAccessFileAdapter());
			com.db4o.Db4o.Configure().Io(adapterFactory);
			com.db4o.ObjectContainer oc = com.db4o.Db4o.OpenFile(fileName);
			com.db4o.ObjectSet objectSet = oc.Get(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase
				(null, "three"));
			oc.Delete(objectSet.Next());
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "four"));
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "five"));
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "six"));
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "seven"));
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "eight"));
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "nine"));
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "10"));
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "11"));
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "12"));
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "13"));
			oc.Set(new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase(null, "14"));
			oc.Commit();
			com.db4o.query.Query q = oc.Query();
			q.Constrain(typeof(com.db4o.db4ounit.common.acid.CrashSimulatingTestCase));
			objectSet = q.Execute();
			while (objectSet.HasNext())
			{
				com.db4o.db4ounit.common.acid.CrashSimulatingTestCase cst = (com.db4o.db4ounit.common.acid.CrashSimulatingTestCase
					)objectSet.Next();
				if (!(cst._name.Equals("10") || cst._name.Equals("13")))
				{
					oc.Delete(cst);
				}
			}
			oc.Commit();
			oc.Close();
			com.db4o.Db4o.Configure().Io(new com.db4o.io.RandomAccessFileAdapter());
			int count = adapterFactory.batch.WriteVersions(fileName);
			CheckFiles(fileName, "R", adapterFactory.batch.NumSyncs());
			CheckFiles(fileName, "W", count);
			j4o.lang.JavaSystem.Out.WriteLine("Total versions: " + count);
		}

		private void CheckFiles(string fileName, string infix, int count)
		{
			for (int i = 1; i <= count; i++)
			{
				string versionedFileName = fileName + infix + i;
				com.db4o.ObjectContainer oc = com.db4o.Db4o.OpenFile(versionedFileName);
				if (!StateBeforeCommit(oc))
				{
					if (!StateAfterFirstCommit(oc))
					{
						Db4oUnit.Assert.IsTrue(StateAfterSecondCommit(oc));
					}
				}
				oc.Close();
			}
		}

		private bool StateBeforeCommit(com.db4o.ObjectContainer oc)
		{
			return Expect(oc, new string[] { "one", "two", "three" });
		}

		private bool StateAfterFirstCommit(com.db4o.ObjectContainer oc)
		{
			return Expect(oc, new string[] { "one", "two", "four", "five", "six", "seven", "eight"
				, "nine", "10", "11", "12", "13", "14" });
		}

		private bool StateAfterSecondCommit(com.db4o.ObjectContainer oc)
		{
			return Expect(oc, new string[] { "10", "13" });
		}

		private bool Expect(com.db4o.ObjectContainer oc, string[] names)
		{
			com.db4o.ObjectSet objectSet = oc.Query(typeof(com.db4o.db4ounit.common.acid.CrashSimulatingTestCase)
				);
			if (objectSet.Size() != names.Length)
			{
				return false;
			}
			while (objectSet.HasNext())
			{
				com.db4o.db4ounit.common.acid.CrashSimulatingTestCase cst = (com.db4o.db4ounit.common.acid.CrashSimulatingTestCase
					)objectSet.Next();
				bool found = false;
				for (int i = 0; i < names.Length; i++)
				{
					if (cst._name.Equals(names[i]))
					{
						names[i] = null;
						found = true;
						break;
					}
				}
				if (!found)
				{
					return false;
				}
			}
			for (int i = 0; i < names.Length; i++)
			{
				if (names[i] != null)
				{
					return false;
				}
			}
			return true;
		}

		private void CreateFile(string fileName)
		{
			com.db4o.ObjectContainer oc = com.db4o.Db4o.OpenFile(fileName);
			for (int i = 0; i < 10; i++)
			{
				oc.Set(new com.db4o.db4ounit.common.assorted.SimplestPossibleItem("delme"));
			}
			com.db4o.db4ounit.common.acid.CrashSimulatingTestCase one = new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase
				(null, "one");
			com.db4o.db4ounit.common.acid.CrashSimulatingTestCase two = new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase
				(one, "two");
			com.db4o.db4ounit.common.acid.CrashSimulatingTestCase three = new com.db4o.db4ounit.common.acid.CrashSimulatingTestCase
				(one, "three");
			oc.Set(one);
			oc.Set(two);
			oc.Set(three);
			oc.Commit();
			com.db4o.ObjectSet objectSet = oc.Query(typeof(com.db4o.db4ounit.common.assorted.SimplestPossibleItem)
				);
			while (objectSet.HasNext())
			{
				oc.Delete(objectSet.Next());
			}
			oc.Close();
			com.db4o.foundation.io.File4.Copy(fileName, fileName + "0");
		}

		public override string ToString()
		{
			return _name + " -> " + _next;
		}
	}
}

File: osgi-projects/db4o_osgi_test/src/com/db4o/db4ounit/common/acid/CrashSimulatingTestCase.java
/* Copyright (C) 2004 - 2006  db4objects Inc.  http://www.db4o.com

This file is part of the db4o open source object database.

db4o is free software; you can redistribute it and/or modify it under
the terms of version 2 of the GNU General Public License as published
by the Free Software Foundation and as clarified by db4objects' GPL 
interpretation policy, available at
http://www.db4o.com/about/company/legalpolicies/gplinterpretation/
Alternatively you can write to db4objects, Inc., 1900 S Norfolk Street,
Suite 350, San Mateo, CA 94403, USA.

db4o is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA  02111-1307, USA. */
package com.db4o.db4ounit.common.acid;

import java.io.*;

import com.db4o.*;
import com.db4o.db4ounit.common.assorted.*;
import com.db4o.foundation.*;
import com.db4o.foundation.io.*;
import com.db4o.internal.*;
import com.db4o.io.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.fixtures.*;


public class CrashSimulatingTestCase implements TestCase, OptOutCS {	
    
    public String _name;
    
    public CrashSimulatingTestCase _next;
    
    
    static final boolean LOG = false;
    
    
    public CrashSimulatingTestCase() {
    }
    
    public CrashSimulatingTestCase(CrashSimulatingTestCase next_, String name) {
        _next = next_;
        _name = name;
    }
    
    private boolean hasLockFileThread(){
        if (!Platform4.hasLockFileThread()) {
            return false;
        }
        return ! Platform4.hasNio();
    }
    
    public void test() throws IOException{
    	if(hasLockFileThread()){
    		System.out.println("CrashSimulatingTestCase is ignored on platforms with lock file thread.");
    		return;
    	}
    	
        String path = Path4.combine(Path4.getTempPath(), "crashSimulate");
        String fileName = Path4.combine(path, "cs");
        
    	File4.delete(fileName);
    	File4.mkdirs(path);
        
    	Db4o.configure().reflectWith(Platform4.reflectorForType(CrashSimulatingTestCase.class));
        Db4o.configure().bTreeNodeSize(4);

        createFile(fileName);
        
        CrashSimulatingIoAdapter adapterFactory = new CrashSimulatingIoAdapter(new RandomAccessFileAdapter());
        Db4o.configure().io(adapterFactory);
        
        ObjectContainer oc = Db4o.openFile(fileName);
        
        ObjectSet objectSet = oc.get(new CrashSimulatingTestCase(null, "three"));
        oc.delete(objectSet.next());
        
        oc.set(new CrashSimulatingTestCase(null, "four"));
        oc.set(new CrashSimulatingTestCase(null, "five"));
        oc.set(new CrashSimulatingTestCase(null, "six"));
        oc.set(new CrashSimulatingTestCase(null, "seven"));
        oc.set(new CrashSimulatingTestCase(null, "eight"));
        oc.set(new CrashSimulatingTestCase(null, "nine"));
        oc.set(new CrashSimulatingTestCase(null, "10"));
        oc.set(new CrashSimulatingTestCase(null, "11"));
        oc.set(new CrashSimulatingTestCase(null, "12"));
        oc.set(new CrashSimulatingTestCase(null, "13"));
        oc.set(new CrashSimulatingTestCase(null, "14"));
        
        oc.commit();
        
        Query q = oc.query();
        q.constrain(CrashSimulatingTestCase.class);
        objectSet = q.execute();
        while(objectSet.hasNext()){
        	CrashSimulatingTestCase cst = (CrashSimulatingTestCase) objectSet.next();
            if( !  (cst._name.equals("10") || cst._name.equals("13")) ){
                oc.delete(cst);
            }
        }
        
        oc.commit();

        oc.close();

        Db4o.configure().io(new RandomAccessFileAdapter());

        int count = adapterFactory.batch.writeVersions(fileName);

        checkFiles(fileName, "R", adapterFactory.batch.numSyncs());
        checkFiles(fileName, "W", count);
		if (LOG) {
			System.out.println("Total versions: " + count);
		}
    }

	private void checkFiles(String fileName, String infix,int count) {
        for (int i = 1; i <= count; i++) {
            if(LOG){
                System.out.println("Checking " + infix + i);
            }
            String versionedFileName = fileName + infix + i;
            ObjectContainer oc = Db4o.openFile(versionedFileName);
	        try {
	            if(! stateBeforeCommit(oc)){
	                if(! stateAfterFirstCommit(oc)){
	                    Assert.isTrue(stateAfterSecondCommit(oc));
	                }
	            }
            } finally {
            	oc.close();
            }
        }
    }
    
    private boolean stateBeforeCommit(ObjectContainer oc){
        return expect(oc, new String[] {"one", "two", "three"});
    }
    
    private boolean stateAfterFirstCommit (ObjectContainer oc){
        return expect(oc, new String[] {"one", "two", "four", "five", "six", "seven", "eight", "nine", "10", "11", "12", "13", "14" });
    }
    
    private boolean stateAfterSecondCommit (ObjectContainer oc){
        return expect(oc, new String[] {"10", "13"});
    }
    
    private boolean expect(ObjectContainer container, String[] names){
    	Collection4 expected = new Collection4(names);
        ObjectSet actual = container.query(CrashSimulatingTestCase.class);
        while (actual.hasNext()){
            CrashSimulatingTestCase current = (CrashSimulatingTestCase)actual.next();
            if (null == expected.remove(current._name)) {
            	return false;
            }
        }
        return expected.isEmpty();
    }
    
    private void createFile(String fileName) throws IOException{
        ObjectContainer oc = Db4o.openFile(fileName);
        try {
        	populate(oc);
        } finally {
        	oc.close();
        }
        File4.copy(fileName, fileName + "0");
    }

	private void populate(ObjectContainer container) {
		for (int i = 0; i < 10; i++) {
            container.set(new SimplestPossibleItem("delme"));
        }
        CrashSimulatingTestCase one = new CrashSimulatingTestCase(null, "one");
        CrashSimulatingTestCase two = new CrashSimulatingTestCase(one, "two");
        CrashSimulatingTestCase three = new CrashSimulatingTestCase(one, "three");
        container.set(one);
        container.set(two);
        container.set(three);
        container.commit();
        ObjectSet objectSet = container.query(SimplestPossibleItem.class);
        while(objectSet.hasNext()){
            container.delete(objectSet.next());
        }
	}
    
	public String toString() {
		return _name+" -> "+_next;
	}
	
}
